### ISSUE TITLE
find_ue function error in tutorial

### ISSUE THREAD

**Issue by Milleupup** [jdlph/Path4GMNS#51]
i'm working on [Tutorial] ,
However,after i upgrade the package,this time find_ue return errors!!!
![image](https://github.com/user-attachments/assets/2850c27e-40d1-4338-9cd6-5550e9ffb59a)

**Comment by jdlph** [comment]
Did you encounter a similar issue when you call find_shortest_path() in [Section 3.1](https://github.com/jdlph/Path4GMNS/blob/master/tutorial/tutorial.ipynb)?

**Comment by Milleupup** [comment]
no,it run successfully:shortest path (node id) from node 1 to node 2, distance: 3.06 mi | node path: 1;547;548;2.shortest path (link id) from node 1 to node 2, distance: 3.06 mi | link path: 1;986;989

**Comment by jdlph** [comment]
Thank you for the update! It helps verify that the C++ path engine works properly on your local machine.

The root cause of your reported issue is also related to it. Specifically, we suspect that the origin_node_no was passed to the path engine as a null pointer. Please provide the following information for us to replicate this issue and conduct further investigation. 

1. Input data.
2. A full screenshot including the code snippet and error traces. 
3. Your OS and build.
4. Python version.

You can directly attach data and images using this thread.

**Comment by Milleupup** [comment]
Thanks! The problem inexplicably disappeared, find_ue returns no errors, but the output route_assignment file is empty?
![image](https://github.com/user-attachments/assets/74b63bb0-d4ff-4c4c-8a3c-1eb96b2ab4e9)
![image](https://github.com/user-attachments/assets/da694ea9-db72-4ad5-b86a-8dfa13d51b6c)
![image](https://github.com/user-attachments/assets/f2758264-ee90-4a5b-a285-4758415d1bde)

**Comment by Milleupup** [comment]
Failed again...
![image](https://github.com/user-attachments/assets/83fe82fc-5ba6-41aa-a180-97d4e2c32764)
![image](https://github.com/user-attachments/assets/e5315c2b-c19a-4074-bc85-480c6f7b6d08)
python:3.11
OS:windows11
data:
![image](https://github.com/user-attachments/assets/421bda8d-a536-4786-8f9c-115833a6c0b0)

**Comment by jdlph** [comment]
Would it be possible for you to zip your two data sets and attach them here so that we are able to replicate these two issues?

**Comment by Milleupup** [comment]
ok, there're my test dataset
[Chicago_Sketch.zip](https://github.com/user-attachments/files/17795327/Chicago_Sketch.zip)

**Comment by jdlph** [comment]
Thank you for sharing your data sets! We are able to replicate these two issues.

The first issue (with respect to the empty route_assignment.csv) is caused by the **data inconsistency on zone id representation** between node.csv and demand.csv. Specifically, zone_id is decimal (e.g., "88.0" in Figure 1) in node.csv but integer in zone.csv (e.g., "1" in Figure 2) even though they are showing as integers in Excel. You can use an advanced text editor (e.g., Notepad++) to verify that. 

This data inconsistency results into no demand being loaded into the UE module. Therefore, you shall have seen zero gaps over the entire column updating process (as illustrated in Figure 3). Empty route_assignment.csv and zero volume for each link in link_performance.csv are expected.

After converting the decimal zone ids into integer values and changing proper units to read_network() (i.e., length_unit='meter', speed_unit='kph'), the issue is gone. Figures 4 and 5 show the terminal output and the first few rows in route_assignment.csv thereafter.

![image](https://github.com/user-attachments/assets/fd1f29d5-f364-47d8-ab05-9a1da0003813)
**Figure 1** Zone id representation in node.csv

![image](https://github.com/user-attachments/assets/e55afe43-05a7-4b1e-b860-c4ff240eafba)
**Figure 2** Zone id representation in demand.csv

![image](https://github.com/user-attachments/assets/b3a28544-4df7-407f-be2c-edb72bb54313)
**Figure 3** Terminal output using the original MyData

![image](https://github.com/user-attachments/assets/932dc7ab-30e5-47d8-b145-ea8f604ce9c1)
**Figure 4** Terminal output after converting decimal zone ids to integers

![image](https://github.com/user-attachments/assets/81abdb12-e3f7-4417-8294-27827fc1eb3c)
**Figure 5** route_assignment.csv after converting decimal zone ids to integers

**Comment by jdlph** [comment]
The first issue regarding OSError is probably due to the execution of find_shortest_path() before find_ue() as shown in your code snippets. The issue was vanished after the invoking of find_shortest_path() is removed or placed after find_ue(). Please the attached screenshots for details.

![image](https://github.com/user-attachments/assets/ec30b3d8-f306-41cc-8249-e8d6f87c6226)
![image](https://github.com/user-attachments/assets/be245e46-6415-4baf-9720-c41d8a1defa5)

The reason behind this issue is elaborated below. 

find_ue() involves a preprocess of adding connectors and centroids to the underlying physical network (add_centroids_connectors()). However, the essential arrays for shortest path calculation are not updated accordingly per the current design of allocate_for_CAPI() in class Network. The column generation process then fails as it tries to find a shortest path tree from a newly added centroid, which is not present in the set of nodes.

This happens to be a **design flaw** as we never thought of invoking find_shortest_path() and find_ue() in the same code snippet. This issue could be prevented from root if we had maintained a virtual network for find_ue() just like what we have done for class AccessNetwork. 

Therefore, the simple and probably the only workaround is to remove the call(s) of find_shortest_path() prior to find_ue().

This issue will **remain active** until we come up with a permanent fix in the source code. Please feel free to add comments and suggestions using this thread.

**Comment by Milleupup** [comment]
thanks a lot, I'll check it later!

**Comment by jdlph** [comment]
[v0.9.9.post1](https://pypi.org/project/path4gmns/0.9.9.post1/) is just released as a hotfix over the reported two issues.

Details regarding the first issue and how it was fixed can be found in the commit message of 2973475. As it still indicates a design flaw for find_ue() under multiple agent types and multiple demand periods, we will leave this ticket open until we permanently fix it.

For the second one, the data inconsistency in terms of zone id representation shall be caught in the first place by throwing the exception below.

```bash
Exception: volume is not encoded as numerical or Every zone id present in demand.csv is not found in node.csv.

For the latter one, different zone id representation could be the reason. For example, zone id is encoded as integer in demand.csv but decimal in node.csv. Note that zone_id CANNOT be decimal per GMNS specification!
Hint: use an advanced text editor (not Excel) to verify.
```

**Comment by jdlph** [comment]
With the extensive use of f-string in Path4GMNS, the minimum runtime is Python 3.6. As the new implementation of dict() introduced in Python 3.6 also preserves the insertion order, the **potential issue of find_ue() under multiple agent types and multiple demand periods** would not exist. More details can be found in [PEP468](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-compactdict) and [PEP498](https://peps.python.org/pep-0498/).

We will update setup.py with this specific Python runtime requirement in the next release.

This issue is now marked as closed. Thank you again for helping us improve Path4GMNS!

**Comment by Milleupup** [comment]
您好，您的邮件我已收到，稍后会给您回复，谢谢！
